UIforETW/PowerStatus.cpp

```diff
@@ -108,125 +108,179 @@ void CPowerStatusMonitor::SampleBatteryStat()
 	for (int deviceNum = 0; deviceNum < maxBatteries; deviceNum++)
 	{
 		SP_DEVICE_INTERFACE_DATA did = { sizeof(did) };
-
-		if (SetupDiEnumDeviceInterfaces(hdev,
+		const BOOL enumDeviceInterfacesResult =
+			SetupDiEnumDeviceInterfaces(hdev,
 										0,
 										&GUID_DEVCLASS_BATTERY,
 										deviceNum,
-										&did))
+										&did);
+		if (!enumDeviceInterfacesResult)
+		{
+			const DWORD lastErr = ::GetLastError( );
+			if (lastErr == ERROR_NO_MORE_ITEMS)
+			{
+				break; // Enumeration failed - perhaps we're out of items
+			}
+			continue;
+		}
+
+		DWORD bytesNeeded = 0;
+		const BOOL getDetailResult = SetupDiGetDeviceInterfaceDetail(hdev,
+										&did,
+										0,
+										0,
+										&bytesNeeded,
+										0);
+
+		if (getDetailResult)
+		{
+			throw std::logic_error("Expected SetupDiGetDeviceInterfaceDetail failure!");
+		}
+
+		if (GetLastError() != ERROR_INSUFFICIENT_BUFFER)
+		{
+			continue;
+		}
+
+
+		HLOCAL detailDataAlloc = LocalAlloc(LPTR, bytesNeeded);

```

Comment:
Heh - no. I started from the MSDN sample code at https://msdn.microsoft.com/en-us/library/windows/desktop/bb204769(v=vs.85).aspx and stopped refactoring once it worked, more or less. I probably should have switched to std::vector<> right then.

A quick check confirms that LocalAlloc has been mostly obsolete since Win32 came into existence 20+ years ago (https://msdn.microsoft.com/en-us/library/windows/desktop/aa366533%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396), so I'm not sure why MSDN samples use it. There are a few places that it is required (clipboard?) but they are rare and this does not appear to be one of them.